app/models/grade_entry_form.rb

```diff
@@ -44,44 +44,56 @@ def out_of_total
 
   # Determine the total mark for a particular student, as a percentage
   def calculate_total_percent(grade_entry_student)
-    unless grade_entry_student.nil?
-      total = grade_entry_student.total_grade
+    unless grade_entry_student.nil? || !@ges_total_grade.nil?
+      @ges_total_grade = grade_entry_student.total_grade
     end
 
     percent = BLANK_MARK
     out_of = self.out_of_total
 
-    # Check for NA mark or division by 0
-    unless total.nil? || out_of == 0
-      percent = (total / out_of) * 100
+    # Check for NA mark f or division by 0
+    unless @ges_total_grade.nil? || out_of == 0
+        percent = (@ges_total_grade / out_of) * 100
     end
+
     percent
   end
 
   # An array of all grade_entry_students' released percentage total grades that are not nil
   def released_percentage_grades_array
     grades = Array.new()
-    grade_entry_students = self.grade_entry_students
-                             .where(released_to_student: true)
-    grade_entry_students.each do |grade_entry_student|
-      if !grade_entry_student.total_grade.nil?
+    ges = GradeEntryStudent.includes(:grades)
+                           .where(released_to_student: true,
+                                  grade_entry_form: self)
+
+    ges.each do |grade_entry_student|
+      @ges_total_grade = grade_entry_student.total_grade
+      unless @ges_total_grade.nil?
         grades.push(calculate_total_percent(grade_entry_student))
       end
     end
 
-    return grades
+    @released_grades_array = grades
```

Comment:
Using an instance variable to store this is a good idea. But you should put a conditional `if defined? @released_grades_array ... return @released_grades_array` at the very top of this method. This is the proper way of caching the result of this method. That way, you can actually keep the exact same code in your `calculate_*` methods that you had before, and the work will only be done the first time `released_percentage_grades_array` is called.